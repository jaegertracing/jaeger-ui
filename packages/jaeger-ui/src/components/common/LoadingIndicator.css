/*
Copyright (c) 2017 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

@keyframes LoadingIndicator--colorAnim {
  /*
  rgb(0, 128, 128) == teal
  rgba(0, 128, 128, 0.3) == #bedfdf
  */
  from {
    color: #bedfdf;
  }
  to {
    color: teal;
  }
}

.LoadingIndicator {
  animation: LoadingIndicator--colorAnim 1s infinite alternate;
  font-size: 36px;
  /* outline / stroke the loading indicator */
  text-shadow: -0.5px 0 rgba(0, 128, 128, 0.6), 0 0.5px rgba(0, 128, 128, 0.6), 0.5px 0 rgba(0, 128, 128, 0.6),
    0 -0.5px rgba(0, 128, 128, 0.6);
}

.LoadingIndicator.is-centered {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.LoadingIndicator.is-vcentered {
  display: block;
  margin: auto;
}

.LoadingIndicator.is-small {
  font-size: 0.7em;
}
