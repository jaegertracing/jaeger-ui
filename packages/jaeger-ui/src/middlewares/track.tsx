// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Action } from 'redux-actions';
import { Dispatch, Store } from 'react-redux';

import { middlewareHooks as searchHooks } from '../components/SearchTracePage/SearchForm.track';
import { middlewareHooks as timelineHooks } from '../components/TracePage/TraceTimelineViewer/duck.track';
import { isWaEnabled } from '../utils/tracking';
import { ReduxState } from '../types';

type TMiddlewareFn = (store: Store<ReduxState>, action: Action<any>) => void;

const middlewareHooks: { [actionType: string]: TMiddlewareFn } = { ...timelineHooks, ...searchHooks };

function trackingMiddleware(store: Store<ReduxState>) {
  return function inner(next: Dispatch<ReduxState>) {
    return function core(action: any) {
      const { type } = action;
      if (typeof middlewareHooks[type] === 'function') {
        middlewareHooks[type](store, action);
      }
      return next(action);
    };
  };
}

export default isWaEnabled ? trackingMiddleware : undefined;
