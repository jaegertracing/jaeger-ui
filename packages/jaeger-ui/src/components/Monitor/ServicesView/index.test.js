// Copyright (c) 2021 The Jaeger Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';
import {
  MonitorATMServicesViewImpl as MonitorATMServicesView,
  mapStateToProps,
  mapDispatchToProps,
  getLoopbackInterval,
  yAxisTickFormat,
  timeFrameOptions,
} from '.';
import LoadingIndicator from '../../common/LoadingIndicator';
import MonitorATMEmptyState from '../EmptyState';
import ServiceGraph from './serviceGraph';
import {
  originInitialState,
  serviceMetrics,
  serviceOpsMetrics,
  serviceMetricsWithOneServiceLatency,
} from '../../../reducers/metrics.mock';
import * as track from './index.track';

const state = {
  services: {},
  metrics: originInitialState,
  selectedService: undefined,
};

const props = mapStateToProps(state);

Date.now = jest.fn(() => 1487076708000); // Tue, 14 Feb 2017 12:51:48 GMT'
jest.mock('lodash/debounce', () => fn => fn);

describe('<MonitorATMServicesView>', () => {
  let wrapper;
  const mockFetchServices = jest.fn();
  const mockFetchAllServiceMetrics = jest.fn();
  const mockFetchAggregatedServiceMetrics = jest.fn();

  beforeAll(() => {
    Date.now = jest.fn(() => 1466424490000);
  });

  beforeEach(() => {
    wrapper = shallow(
      <MonitorATMServicesView
        {...props}
        fetchServices={mockFetchServices}
        fetchAllServiceMetrics={mockFetchAllServiceMetrics}
        fetchAggregatedServiceMetrics={mockFetchAggregatedServiceMetrics}
      />
    );
  });

  afterEach(() => {
    wrapper = null;

    jest.clearAllMocks();
  });

  it('does not explode', () => {
    expect(wrapper.length).toBe(1);
  });

  it('shows a loading indicator when loading services list', () => {
    wrapper.setProps({ servicesLoading: true });
    expect(wrapper.find(LoadingIndicator).length).toBe(1);
  });

  it('do not show a loading indicator once data loaded', () => {
    wrapper.setProps({
      services: ['s1'],
      selectedService: undefined,
      metrics: {
        ...originInitialState,
        serviceMetrics,
        serviceOpsMetrics,
        loading: false,
        isATMActivated: true,
      },
    });
    expect(wrapper.find(LoadingIndicator).length).toBe(0);
  });

  it('render one service latency', () => {
    wrapper.setProps({
      services: ['s1'],
      selectedService: undefined,
      metrics: {
        ...originInitialState,
        serviceMetricsWithOneServiceLatency,
        serviceOpsMetrics,
        loading: false,
        isATMActivated: true,
      },
    });
    expect(wrapper.length).toBe(1);
  });

  it('Render ATM not configured page', () => {
    wrapper.setProps({
      services: [],
      selectedService: undefined,
      metrics: {
        ...originInitialState,
        serviceMetrics,
        serviceOpsMetrics,
        loading: false,
        isATMActivated: false,
      },
    });
    expect(wrapper.find(MonitorATMEmptyState).length).toBe(1);
  });

  it('function invocation check on page load', () => {
    expect(mockFetchServices).toHaveBeenCalled();
    expect(mockFetchAllServiceMetrics).not.toHaveBeenCalled();
    expect(mockFetchAggregatedServiceMetrics).not.toHaveBeenCalled();
    wrapper.setProps({
      services: ['s1'],
      metrics: {
        ...originInitialState,
        serviceMetrics,
        serviceOpsMetrics,
        loading: false,
        isATMActivated: true,
      },
    });
    expect(mockFetchAllServiceMetrics).toHaveBeenCalled();
    expect(mockFetchAggregatedServiceMetrics).toHaveBeenCalled();
  });

  it('ATM snapshot test', () => {
    mockFetchServices.mockResolvedValue(['s1', 's2']);
    wrapper.setProps({
      services: ['s1', 's2'],
      metrics: {
        ...originInitialState,
        serviceMetrics,
        serviceOpsMetrics,
        loading: false,
        isATMActivated: true,
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('ATM snapshot test with no metrics', () => {
    mockFetchServices.mockResolvedValue([]);
    wrapper.setProps({
      metrics: {
        ...originInitialState,
        serviceMetrics: {
          service_latencies: null,
        },
        serviceOpsMetrics,
        loading: false,
        isATMActivated: true,
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('render one service latency', () => {
    wrapper.setProps({
      metrics: {
        ...originInitialState,
        serviceMetricsWithOneServiceLatency,
        serviceOpsMetrics,
        loading: false,
        isATMActivated: true,
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('ComponentWillUnmount remove listener', () => {
    const remover = jest.spyOn(global, 'removeEventListener').mockImplementation(() => {});
    wrapper.unmount();
    expect(remover).toHaveBeenCalled();
  });

  it('resize window test', () => {
    const selectedInput = 'graphDivWrapper';
    wrapper.instance()[selectedInput] = {
      current: {
        offsetWidth: 100,
      },
    };
    global.dispatchEvent(new Event('resize'));
    expect(wrapper.state().graphWidth).toBe(76);
  });

  it('should update state after choosing a new timeframe', () => {
    const firstGraphXDomain = wrapper.state().graphXDomain;
    wrapper.setProps({
      selectedTimeFrame: 3600000 * 2,
    });

    expect(wrapper.state().graphXDomain).not.toBe(firstGraphXDomain);
  });

  it('search test', () => {
    mockFetchServices.mockResolvedValue(['cartservice']);
    wrapper.setProps({
      services: ['s1', 's2'],
      metrics: {
        ...originInitialState,
        serviceMetrics,
        serviceOpsMetrics,
        loading: false,
        isATMActivated: true,
      },
    });

    wrapper.find('Search').simulate('change', { target: { value: 'place' } });
    expect(wrapper.state().serviceOpsMetrics.length).toBe(1);
    wrapper.find('Search').simulate('change', { target: { value: 'qqq' } });
    expect(wrapper.state().serviceOpsMetrics.length).toBe(0);
    wrapper.find('Search').simulate('change', { target: { value: '' } });
    expect(wrapper.state().serviceOpsMetrics.length).toBe(1);
  });

  it('Error in serviceLatencies ', () => {
    wrapper.setProps({
      services: ['s1', 's2'],
      selectedService: 's1',
      metrics: {
        ...originInitialState,
        serviceMetrics,
        serviceOpsMetrics,
        loading: false,
        isATMActivated: true,
        serviceError: {
          ...originInitialState.serviceError,
          service_latencies_50: new Error('some API error'),
        },
      },
    });
    expect(
      wrapper
        .find(ServiceGraph)
        .first()
        .prop('error')
    ).toBeNull();

    wrapper.setProps({
      services: ['s1', 's2'],
      selectedService: 's1',
      metrics: {
        ...originInitialState,
        serviceMetrics,
        serviceOpsMetrics,
        loading: false,
        isATMActivated: true,
        serviceError: {
          ...originInitialState.serviceError,
          service_latencies_50: new Error('some API error'),
          service_latencies_75: new Error('some API error'),
        },
      },
    });
    expect(
      wrapper
        .find(ServiceGraph)
        .first()
        .prop('error')
    ).toBeNull();

    wrapper.setProps({
      services: ['s1', 's2'],
      selectedService: 's1',
      metrics: {
        ...originInitialState,
        serviceMetrics,
        serviceOpsMetrics,
        loading: false,
        isATMActivated: true,
        serviceError: {
          service_latencies_50: new Error('some API error'),
          service_latencies_75: new Error('some API error'),
          service_latencies_95: new Error('some API error'),
        },
      },
    });
    expect(
      wrapper
        .find(ServiceGraph)
        .first()
        .prop('error')
    ).not.toBeNull();
  });

  it('Should track all events', () => {
    const trackSelectServiceSpy = jest.spyOn(track, 'trackSelectService');
    const trackViewAllTracesSpy = jest.spyOn(track, 'trackViewAllTraces');
    const trackSelectTimeframeSpy = jest.spyOn(track, 'trackSelectTimeframe');
    const trackSearchOperationSpy = jest.spyOn(track, 'trackSearchOperation');

    const newValue = 'newValue';
    const [timeFrameOption] = timeFrameOptions;

    wrapper.setProps({
      metrics: { ...originInitialState, serviceOpsMetrics },
    });

    wrapper.find('Search').simulate('change', { target: { value: newValue } });
    expect(trackSearchOperationSpy).toHaveBeenCalledWith(newValue);

    wrapper
      .find('Field')
      .first()
      .simulate('change', null, newValue);
    expect(trackSelectServiceSpy).toHaveBeenCalledWith(newValue);

    wrapper
      .find('Field')
      .last()
      .simulate('change', null, timeFrameOption.value);
    expect(trackSelectTimeframeSpy).toHaveBeenCalledWith(timeFrameOption.label);

    wrapper.find({ children: 'View all traces' }).simulate('click');
    expect(trackViewAllTracesSpy).toHaveBeenCalled();

    trackSelectServiceSpy.mockReset();
    trackViewAllTracesSpy.mockReset();
    trackSelectTimeframeSpy.mockReset();
    trackSearchOperationSpy.mockReset();
  });
});

describe('<MonitorATMServicesView> on page switch', () => {
  // eslint-disable-next-line no-unused-vars
  let wrapper;
  const stateOnPageSwitch = {
    services: {
      services: ['s1'],
    },
    metrics: originInitialState,
    selectedService: undefined,
  };

  const propsOnPageSwitch = mapStateToProps(stateOnPageSwitch);
  const mockFetchServices = jest.fn();
  const mockFetchAllServiceMetrics = jest.fn();
  const mockFetchAggregatedServiceMetrics = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <MonitorATMServicesView
        {...propsOnPageSwitch}
        fetchServices={mockFetchServices}
        fetchAllServiceMetrics={mockFetchAllServiceMetrics}
        fetchAggregatedServiceMetrics={mockFetchAggregatedServiceMetrics}
      />
    );
  });

  it('function invocation check on page load', () => {
    expect(mockFetchServices).toHaveBeenCalled();
    expect(mockFetchAllServiceMetrics).toHaveBeenCalled();
    expect(mockFetchAggregatedServiceMetrics).toHaveBeenCalled();
  });
});

describe('mapStateToProps()', () => {
  it('refines state to generate the props', () => {
    expect(mapStateToProps(state)).toEqual({
      metrics: originInitialState,
      services: [],
      selectedService: 's1',
      selectedTimeFrame: 3600000,
    });
  });
});

describe('mapDispatchToProps()', () => {
  it('providers the `fetchServices` , `fetchAllServiceMetrics` and `fetchAggregatedServiceMetrics` prop', () => {
    expect(mapDispatchToProps({})).toEqual({
      fetchServices: expect.any(Function),
      fetchAllServiceMetrics: expect.any(Function),
      fetchAggregatedServiceMetrics: expect.any(Function),
    });
  });
});

describe('getLoopbackInterval()', () => {
  it('undefined value', () => {
    expect(getLoopbackInterval()).toBe('');
  });

  it('timeframe NOT exists', () => {
    expect(getLoopbackInterval(111)).toBe('');
  });

  it('timeframe exists', () => {
    expect(getLoopbackInterval(48 * 3600000)).toBe('last 2 days');
  });
});

describe('yAxisTickFormat', () => {
  it('value is 2 second', () => {
    const timeInMs = 1000;
    const displayTimeUnit = 'seconds';
    expect(yAxisTickFormat(timeInMs, displayTimeUnit)).toBe(1);
  });
});
