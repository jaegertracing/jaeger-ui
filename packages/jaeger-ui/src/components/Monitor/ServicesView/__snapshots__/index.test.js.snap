// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MonitorATMServicesView> ATM snapshot test 1`] = `
<Fragment>
  <div
    className="service-view-container"
  >
    <Row
      gutter={0}
    >
      <Col
        span={6}
      >
        <h2
          className="service-selector-header"
        >
          Choose service
        </h2>
        <Field
          component={[Function]}
          name="service"
          onChange={[Function]}
          placeholder="Select A Service"
          props={
            Object {
              "className": "select-a-service-input",
              "clearable": false,
              "disabled": null,
              "options": Array [
                Object {
                  "label": "s1",
                  "value": "s1",
                },
                Object {
                  "label": "s2",
                  "value": "s2",
                },
              ],
              "required": true,
              "value": "s1",
            }
          }
        />
      </Col>
    </Row>
    <Row
      gutter={0}
    >
      <Col
        span={16}
      >
        <p
          className="operations-metrics-text"
        >
          Aggregation of all "
          s1
          " metrics in selected timeframe.
           
          <a
            href="/search?end=1466424490000000&limit=20&lookback=1h&maxDuration&minDuration&service=s1&start=1466420890000000"
            onClick={[Function]}
            target="blank"
          >
            View all traces
          </a>
        </p>
      </Col>
      <Col
        className="timeframe-selector"
        span={8}
      >
        <Field
          component={[Function]}
          name="timeframe"
          onChange={[Function]}
          placeholder="Select A Timeframe"
          props={
            Object {
              "className": "select-a-timeframe-input",
              "clearable": false,
              "defaultValue": Object {
                "label": "Last Hour",
                "value": 3600000,
              },
              "disabled": null,
              "options": Array [
                Object {
                  "label": "Last 5 minutes",
                  "value": 300000,
                },
                Object {
                  "label": "Last 15 minutes",
                  "value": 900000,
                },
                Object {
                  "label": "Last 30 minutes",
                  "value": 1800000,
                },
                Object {
                  "label": "Last Hour",
                  "value": 3600000,
                },
                Object {
                  "label": "Last 2 hours",
                  "value": 7200000,
                },
                Object {
                  "label": "Last 6 hours",
                  "value": 21600000,
                },
                Object {
                  "label": "Last 12 hours",
                  "value": 43200000,
                },
                Object {
                  "label": "Last 24 hours",
                  "value": 86400000,
                },
                Object {
                  "label": "Last 2 days",
                  "value": 172800000,
                },
              ],
              "required": true,
              "value": 3600000,
            }
          }
        />
      </Col>
    </Row>
    <Row
      gutter={0}
    >
      <Col
        span={8}
      >
        <div />
        <ServiceGraphImpl
          error={null}
          key="latency"
          loading={false}
          marginClassName="latency-margins"
          metricsData={
            Array [
              Object {
                "max": 189.86,
                "metricPoints": Array [
                  Object {
                    "x": 1631271823806,
                    "y": 189.86,
                  },
                  Object {
                    "x": 1631271883806,
                    "y": 189.86,
                  },
                ],
                "quantile": 0.5,
                "serviceName": "cartservice",
              },
              Object {
                "max": 189.86,
                "metricPoints": Array [
                  Object {
                    "x": 1631271823806,
                    "y": 189.86,
                  },
                  Object {
                    "x": 1631271883806,
                    "y": 189.86,
                  },
                ],
                "quantile": 0.75,
                "serviceName": "cartservice",
              },
              Object {
                "max": 189.86,
                "metricPoints": Array [
                  Object {
                    "x": 1631271823806,
                    "y": 189.86,
                  },
                  Object {
                    "x": 1631271883806,
                    "y": 189.86,
                  },
                ],
                "quantile": 0.95,
                "serviceName": "cartservice",
              },
            ]
          }
          name="Latency (ms)"
          showHorizontalLines={true}
          showLegend={true}
          width={300}
          xDomain={
            Array [
              1466420890000,
              1466424490000,
            ]
          }
          yAxisTickFormat={[Function]}
        />
      </Col>
      <Col
        span={8}
      >
        <ServiceGraphImpl
          color="#CD513A"
          error={null}
          key="errRate"
          loading={false}
          marginClassName="error-rate-margins"
          metricsData={
            Object {
              "max": 1,
              "metricPoints": Array [
                Object {
                  "x": 1631274747520,
                  "y": 100,
                },
                Object {
                  "x": 1631274807520,
                  "y": 100,
                },
              ],
              "quantile": 0.95,
              "serviceName": "cartservice",
            }
          }
          name="Error rate (%)"
          width={300}
          xDomain={
            Array [
              1466420890000,
              1466424490000,
            ]
          }
          yDomain={
            Array [
              0,
              100,
            ]
          }
        />
      </Col>
      <Col
        span={8}
      >
        <ServiceGraphImpl
          color="#4795BA"
          error={null}
          key="requests"
          loading={false}
          marginClassName="request-margins"
          metricsData={
            Object {
              "max": 0.05,
              "metricPoints": Array [
                Object {
                  "x": 1631271823806,
                  "y": 0.05,
                },
                Object {
                  "x": 1631271883806,
                  "y": 0.05,
                },
              ],
              "quantile": 0.95,
              "serviceName": "cartservice",
            }
          }
          name="Request rate (req/s)"
          showHorizontalLines={true}
          width={300}
          xDomain={
            Array [
              1466420890000,
              1466424490000,
            ]
          }
        />
      </Col>
    </Row>
    <Row
      className="operation-table-block"
      gutter={0}
    >
      <Row
        gutter={0}
      >
        <Col
          span={16}
        >
          <h2
            className="table-header"
          >
            Operations metrics under 
            s1
          </h2>
           
          <span
            className="over-the-last"
          >
            Over the 
            last hour
          </span>
        </Col>
        <Col
          className="select-operation-column"
          span={8}
        >
          <Search
            className="select-operation-input"
            disabled={false}
            enterButton={false}
            inputPrefixCls="ant-input"
            onChange={[Function]}
            placeholder="Search operation"
            prefixCls="ant-input-search"
            value=""
          />
        </Col>
      </Row>
      <Row
        gutter={0}
      >
        <OperationTableDetails
          data={
            Array [
              Object {
                "dataPoints": Object {
                  "avg": Object {
                    "service_operation_call_rate": 0.01,
                    "service_operation_error_rate": 1,
                    "service_operation_latencies": 736.16,
                  },
                  "service_operation_call_rate": Array [
                    Object {
                      "x": 1631534436235,
                      "y": 0.01,
                    },
                    Object {
                      "x": 1631534496235,
                      "y": 0.01,
                    },
                  ],
                  "service_operation_error_rate": Array [
                    Object {
                      "x": 1631534436235,
                      "y": 1,
                    },
                    Object {
                      "x": 1631534496235,
                      "y": 1,
                    },
                  ],
                  "service_operation_latencies": Array [
                    Object {
                      "x": 1631534436235,
                      "y": 737.33,
                    },
                    Object {
                      "x": 1631534496235,
                      "y": 735,
                    },
                  ],
                },
                "errRates": 1,
                "impact": 1,
                "key": 0,
                "latency": 736.16,
                "name": "/PlaceOrder",
                "requests": 0.01,
              },
            ]
          }
          endTime={1466424490000}
          error={
            Object {
              "opsCalls": null,
              "opsErrors": null,
              "opsLatencies": null,
            }
          }
          loading={null}
          lookback={3600000}
          serviceName="s1"
        />
      </Row>
    </Row>
  </div>
</Fragment>
`;

exports[`<MonitorATMServicesView> ATM snapshot test with no metrics 1`] = `
<Fragment>
  <Alert
    message={
      <React.Fragment>
        No data yet! Please see these
        <Link
          replace={false}
          target="_blank"
          to={
            Object {
              "pathname": "https://www.jaegertracing.io/docs/latest/spm/",
            }
          }
        >
           instructions 
        </Link>
        on how to set up your span metrics.
      </React.Fragment>
    }
    showIcon={true}
    type="warning"
  />
  <div
    className="service-view-container"
  >
    <Row
      gutter={0}
    >
      <Col
        span={6}
      >
        <h2
          className="service-selector-header"
        >
          Choose service
        </h2>
        <Field
          component={[Function]}
          name="service"
          onChange={[Function]}
          placeholder="Select A Service"
          props={
            Object {
              "className": "select-a-service-input",
              "clearable": false,
              "disabled": null,
              "options": Array [],
              "required": true,
              "value": "s1",
            }
          }
        />
      </Col>
    </Row>
    <Row
      gutter={0}
    >
      <Col
        span={16}
      >
        <p
          className="operations-metrics-text"
        >
          Aggregation of all "
          s1
          " metrics in selected timeframe.
           
          <a
            href="/search?end=1466424490000000&limit=20&lookback=1h&maxDuration&minDuration&service=s1&start=1466420890000000"
            onClick={[Function]}
            target="blank"
          >
            View all traces
          </a>
        </p>
      </Col>
      <Col
        className="timeframe-selector"
        span={8}
      >
        <Field
          component={[Function]}
          name="timeframe"
          onChange={[Function]}
          placeholder="Select A Timeframe"
          props={
            Object {
              "className": "select-a-timeframe-input",
              "clearable": false,
              "defaultValue": Object {
                "label": "Last Hour",
                "value": 3600000,
              },
              "disabled": null,
              "options": Array [
                Object {
                  "label": "Last 5 minutes",
                  "value": 300000,
                },
                Object {
                  "label": "Last 15 minutes",
                  "value": 900000,
                },
                Object {
                  "label": "Last 30 minutes",
                  "value": 1800000,
                },
                Object {
                  "label": "Last Hour",
                  "value": 3600000,
                },
                Object {
                  "label": "Last 2 hours",
                  "value": 7200000,
                },
                Object {
                  "label": "Last 6 hours",
                  "value": 21600000,
                },
                Object {
                  "label": "Last 12 hours",
                  "value": 43200000,
                },
                Object {
                  "label": "Last 24 hours",
                  "value": 86400000,
                },
                Object {
                  "label": "Last 2 days",
                  "value": 172800000,
                },
              ],
              "required": true,
              "value": 3600000,
            }
          }
        />
      </Col>
    </Row>
    <Row
      gutter={0}
    >
      <Col
        span={8}
      >
        <div />
        <ServiceGraphImpl
          error={null}
          key="latency"
          loading={false}
          marginClassName="latency-margins"
          metricsData={null}
          name="Latency (μs)"
          showHorizontalLines={true}
          showLegend={true}
          width={300}
          xDomain={
            Array [
              1466420890000,
              1466424490000,
            ]
          }
          yAxisTickFormat={[Function]}
        />
      </Col>
      <Col
        span={8}
      >
        <ServiceGraphImpl
          color="#CD513A"
          error={null}
          key="errRate"
          loading={false}
          marginClassName="error-rate-margins"
          metricsData={null}
          name="Error rate (%)"
          width={300}
          xDomain={
            Array [
              1466420890000,
              1466424490000,
            ]
          }
          yDomain={
            Array [
              0,
              100,
            ]
          }
        />
      </Col>
      <Col
        span={8}
      >
        <ServiceGraphImpl
          color="#4795BA"
          error={null}
          key="requests"
          loading={false}
          marginClassName="request-margins"
          name="Request rate (req/s)"
          showHorizontalLines={true}
          width={300}
          xDomain={
            Array [
              1466420890000,
              1466424490000,
            ]
          }
        />
      </Col>
    </Row>
    <Row
      className="operation-table-block"
      gutter={0}
    >
      <Row
        gutter={0}
      >
        <Col
          span={16}
        >
          <h2
            className="table-header"
          >
            Operations metrics under 
            s1
          </h2>
           
          <span
            className="over-the-last"
          >
            Over the 
            last hour
          </span>
        </Col>
        <Col
          className="select-operation-column"
          span={8}
        >
          <Search
            className="select-operation-input"
            disabled={false}
            enterButton={false}
            inputPrefixCls="ant-input"
            onChange={[Function]}
            placeholder="Search operation"
            prefixCls="ant-input-search"
            value=""
          />
        </Col>
      </Row>
      <Row
        gutter={0}
      >
        <OperationTableDetails
          data={
            Array [
              Object {
                "dataPoints": Object {
                  "avg": Object {
                    "service_operation_call_rate": 0.01,
                    "service_operation_error_rate": 1,
                    "service_operation_latencies": 736.16,
                  },
                  "service_operation_call_rate": Array [
                    Object {
                      "x": 1631534436235,
                      "y": 0.01,
                    },
                    Object {
                      "x": 1631534496235,
                      "y": 0.01,
                    },
                  ],
                  "service_operation_error_rate": Array [
                    Object {
                      "x": 1631534436235,
                      "y": 1,
                    },
                    Object {
                      "x": 1631534496235,
                      "y": 1,
                    },
                  ],
                  "service_operation_latencies": Array [
                    Object {
                      "x": 1631534436235,
                      "y": 737.33,
                    },
                    Object {
                      "x": 1631534496235,
                      "y": 735,
                    },
                  ],
                },
                "errRates": 1,
                "impact": 1,
                "key": 0,
                "latency": 736.16,
                "name": "/PlaceOrder",
                "requests": 0.01,
              },
            ]
          }
          endTime={1466424490000}
          error={
            Object {
              "opsCalls": null,
              "opsErrors": null,
              "opsLatencies": null,
            }
          }
          loading={null}
          lookback={3600000}
          serviceName="s1"
        />
      </Row>
    </Row>
  </div>
</Fragment>
`;

exports[`<MonitorATMServicesView> render one service latency 1`] = `
<Fragment>
  <Alert
    message={
      <React.Fragment>
        No data yet! Please see these
        <Link
          replace={false}
          target="_blank"
          to={
            Object {
              "pathname": "https://www.jaegertracing.io/docs/latest/spm/",
            }
          }
        >
           instructions 
        </Link>
        on how to set up your span metrics.
      </React.Fragment>
    }
    showIcon={true}
    type="warning"
  />
  <div
    className="service-view-container"
  >
    <Row
      gutter={0}
    >
      <Col
        span={6}
      >
        <h2
          className="service-selector-header"
        >
          Choose service
        </h2>
        <Field
          component={[Function]}
          name="service"
          onChange={[Function]}
          placeholder="Select A Service"
          props={
            Object {
              "className": "select-a-service-input",
              "clearable": false,
              "disabled": null,
              "options": Array [],
              "required": true,
              "value": "s1",
            }
          }
        />
      </Col>
    </Row>
    <Row
      gutter={0}
    >
      <Col
        span={16}
      >
        <p
          className="operations-metrics-text"
        >
          Aggregation of all "
          s1
          " metrics in selected timeframe.
           
          <a
            href="/search?end=1466424490000000&limit=20&lookback=1h&maxDuration&minDuration&service=s1&start=1466420890000000"
            onClick={[Function]}
            target="blank"
          >
            View all traces
          </a>
        </p>
      </Col>
      <Col
        className="timeframe-selector"
        span={8}
      >
        <Field
          component={[Function]}
          name="timeframe"
          onChange={[Function]}
          placeholder="Select A Timeframe"
          props={
            Object {
              "className": "select-a-timeframe-input",
              "clearable": false,
              "defaultValue": Object {
                "label": "Last Hour",
                "value": 3600000,
              },
              "disabled": null,
              "options": Array [
                Object {
                  "label": "Last 5 minutes",
                  "value": 300000,
                },
                Object {
                  "label": "Last 15 minutes",
                  "value": 900000,
                },
                Object {
                  "label": "Last 30 minutes",
                  "value": 1800000,
                },
                Object {
                  "label": "Last Hour",
                  "value": 3600000,
                },
                Object {
                  "label": "Last 2 hours",
                  "value": 7200000,
                },
                Object {
                  "label": "Last 6 hours",
                  "value": 21600000,
                },
                Object {
                  "label": "Last 12 hours",
                  "value": 43200000,
                },
                Object {
                  "label": "Last 24 hours",
                  "value": 86400000,
                },
                Object {
                  "label": "Last 2 days",
                  "value": 172800000,
                },
              ],
              "required": true,
              "value": 3600000,
            }
          }
        />
      </Col>
    </Row>
    <Row
      gutter={0}
    >
      <Col
        span={8}
      >
        <div />
        <ServiceGraphImpl
          error={null}
          key="latency"
          loading={false}
          marginClassName="latency-margins"
          metricsData={null}
          name="Latency (μs)"
          showHorizontalLines={true}
          showLegend={true}
          width={300}
          xDomain={
            Array [
              1466420890000,
              1466424490000,
            ]
          }
          yAxisTickFormat={[Function]}
        />
      </Col>
      <Col
        span={8}
      >
        <ServiceGraphImpl
          color="#CD513A"
          error={null}
          key="errRate"
          loading={false}
          marginClassName="error-rate-margins"
          metricsData={null}
          name="Error rate (%)"
          width={300}
          xDomain={
            Array [
              1466420890000,
              1466424490000,
            ]
          }
          yDomain={
            Array [
              0,
              100,
            ]
          }
        />
      </Col>
      <Col
        span={8}
      >
        <ServiceGraphImpl
          color="#4795BA"
          error={null}
          key="requests"
          loading={false}
          marginClassName="request-margins"
          metricsData={null}
          name="Request rate (req/s)"
          showHorizontalLines={true}
          width={300}
          xDomain={
            Array [
              1466420890000,
              1466424490000,
            ]
          }
        />
      </Col>
    </Row>
    <Row
      className="operation-table-block"
      gutter={0}
    >
      <Row
        gutter={0}
      >
        <Col
          span={16}
        >
          <h2
            className="table-header"
          >
            Operations metrics under 
            s1
          </h2>
           
          <span
            className="over-the-last"
          >
            Over the 
            last hour
          </span>
        </Col>
        <Col
          className="select-operation-column"
          span={8}
        >
          <Search
            className="select-operation-input"
            disabled={false}
            enterButton={false}
            inputPrefixCls="ant-input"
            onChange={[Function]}
            placeholder="Search operation"
            prefixCls="ant-input-search"
            value=""
          />
        </Col>
      </Row>
      <Row
        gutter={0}
      >
        <OperationTableDetails
          data={
            Array [
              Object {
                "dataPoints": Object {
                  "avg": Object {
                    "service_operation_call_rate": 0.01,
                    "service_operation_error_rate": 1,
                    "service_operation_latencies": 736.16,
                  },
                  "service_operation_call_rate": Array [
                    Object {
                      "x": 1631534436235,
                      "y": 0.01,
                    },
                    Object {
                      "x": 1631534496235,
                      "y": 0.01,
                    },
                  ],
                  "service_operation_error_rate": Array [
                    Object {
                      "x": 1631534436235,
                      "y": 1,
                    },
                    Object {
                      "x": 1631534496235,
                      "y": 1,
                    },
                  ],
                  "service_operation_latencies": Array [
                    Object {
                      "x": 1631534436235,
                      "y": 737.33,
                    },
                    Object {
                      "x": 1631534496235,
                      "y": 735,
                    },
                  ],
                },
                "errRates": 1,
                "impact": 1,
                "key": 0,
                "latency": 736.16,
                "name": "/PlaceOrder",
                "requests": 0.01,
              },
            ]
          }
          endTime={1466424490000}
          error={
            Object {
              "opsCalls": null,
              "opsErrors": null,
              "opsLatencies": null,
            }
          }
          loading={null}
          lookback={3600000}
          serviceName="s1"
        />
      </Row>
    </Row>
  </div>
</Fragment>
`;
